### ISSUE TITLE
non_ascii_literal should not be machine-applicable to raw strings

### ISSUE THREAD

**Issue by alexjago** [rust-lang/rust-clippy#9805]
### Summary

`non_ascii_literal` is machine-applied by substituting unicode escapes. However, those are incompatible with raw strings.

(And if like me you try using raw strings as formatting targets, you'll then get errors about invalid references to positional arguments.)

### Reproducer

This code should trigger the lint:

```rust
format!(r#"a raw string with a range {}–{} (and an en dash)"#, 31, 0);
```

`cargo clippy --fix` reported the following:

```
The following errors were reported:
error: invalid reference to positional argument 2013 (there are (...) arguments)
   --> src/filename.rs:(line):(col)
    |
line |         format!(r#"a raw string with a range {}\u{2013}{} (and an en dash)"#, 31, 0);
```

### Version

```text
rustc 1.62.1 (e092d0b6b 2022-07-16)
binary: rustc
commit-hash: e092d0b6b43f2de967af0887873151bb1c0b18d3
commit-date: 2022-07-16
host: x86_64-apple-darwin
release: 1.62.1
LLVM version: 14.0.5
```

### Additional Labels

@rustbot +label I-suggestion-causes-error I-false-positive

**Comment by alexjago** [comment]
That'd be the easiest fix but there's a few things going on as I see them:

First is whether or not to _report_ for raw strings - maybe someone is trying to enforce an all-ASCII codebase with this lint. 

Second, if it _is_ still reported, there _could_ be a hint along the lines of "Clippy can replace non-ASCII characters with substitutions outside of raw strings". 

I suppose there's also a potential restriction lint which would convert raw strings to normal strings - in which case a subsequent fix could apply the unicode substitution. 

I went looking and wow, this goes back to #85 and #299, looks like the original basis of this lint was trying to do something about zero-width spaces and visually similar unicode characters.

**Comment by alexjago** [comment]
I've been having a read through https://github.com/rust-lang/rust-clippy/blob/master/clippy_lints/src/unicode.rs

The lint `invisible_characters` experiences the same issue (machine-applied fix on raw strings), except that it's deny-by-default. 

As for `unicode_not_nfc`, it's OK (i.e. not escaped) without `non_ascii_literal` . (If both are active, then the suggested NAL replacement is in NFC form - this is fine too.)

I've got a minimal repo up here: https://github.com/alexjago/clippy_unicode_tests

**Comment by tylerjw** [comment]
I tried getting started on working on this and wrote this function, used your examples in the testfile `unicode.rs`, however it doesn't seem to work.

```rust
fn is_raw_str(expr: &'_ Expr<'_>) -> bool {
    if let ExprKind::Lit(ref lit) = expr.kind {
        if let LitKind::Str(_, style) = lit.node {
            if matches!(style, StrStyle::Raw(_)) {
                return true;
            }
        }
    }
    false
}
```

Do you know why this function might not work on the example raw string literals?

**Comment by xFrednet** [comment]
Hmmm, I'm sadly not an expert, when it comes to the `format!` macro. You could try attaching a `#[clippy::dump]` attribute (then run Clippy) and look at the output.

**Comment by tylerjw** [comment]
Your intuition is good here. Something about the `format!` macro splits this string into an Array object of `Cooked` string literals (not raw strings). Here is part of the output from `#[clippy::dump]`:  [playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e715a76a6363ccc0a2e7b0df52ca618c)

```
kind: Array(
    [
        Expr {
            hir_id: HirId(DefId(0:3 ~ playground[89a1]::main).10),
            kind: Lit(
                Spanned {
                    node: Str(
                        "a raw string with a range ",
                        Cooked,
                    ),
                    span: src/main.rs:5:21: 5:74 (#0),
                },
            ),
            span: src/main.rs:5:21: 5:74 (#0),
        },
        Expr {
            hir_id: HirId(DefId(0:3 ~ playground[89a1]::main).11),
            kind: Lit(
                Spanned {
                    node: Str(
                        "–",
                        Cooked,
                    ),
                    span: src/main.rs:5:21: 5:74 (#0),
                },
            ),
            span: src/main.rs:5:21: 5:74 (#0),
        },
        Expr {
            hir_id: HirId(DefId(0:3 ~ playground[89a1]::main).12),
            kind: Lit(
                Spanned {
                    node: Str(
                        " (and an en dash)",
                        Cooked,
                    ),
                    span: src/main.rs:5:21: 5:74 (#0),
                },
            ),
            span: src/main.rs:5:21: 5:74 (#0),
        },
    ],
),
```

**Comment by xFrednet** [comment]
Hey, @nyurik and @Alexendoo, I believe you're the best `format!` macro wizards in Clippy rn. Do you maybe have an idea how to fix this issue? Since the strings are `Cooked`, I'm not sure we can find a better way, then snipping the code and checking for a `r#*"` snippet :thinking:

**Comment by nyurik** [comment]
I'm more of an apprentice :)  Note that just a few days ago https://github.com/rust-lang/rust/pull/106745 was merged - which will require a significant rework of all the format-related lint code, so the above might no longer be relevant...

**Comment by Alexendoo** [comment]
Tricky one, It looks like the lint does take a snippet to examine already, so checking if that starts with `r` would be a good way to bypass the issue of them being `Cooked` in the HIR

The suggestion to still lint raw strings but change the suggestion to mention it'd have to be converted to a normal string sounds good to me

**Comment by ada4a** [comment]
> The suggestion to still lint raw strings but change the suggestion to mention it'd have to be converted to a normal string sounds good to me

I tried implementing that, and got the following diagnostic:
```
error: literal non-ASCII character detected
  --> tests/ui/unicode_unfixable.rs:6:21
   |
LL |     let a = format!(r#"a raw string with a range {}–{} (and an en dash)"#, 31, 0);
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider replacing the string with: `r#"a raw string with a range {}\u{2013}{} (and an en dash)"#`
   |
   = help: you'll need to turn the raw string literal into a regular literal first
```
(no idea why the whole suggestion is still given on the same line btw -- it's clearly too long to be legible)

But I think it's rather confusing to suggest a raw literal and then say that the user actually needs to have the literal non-raw.

I thought about borrowing the raw-to-regular conversion logic from `raw_strings`, but it looks a bit complicated...

**Comment by Alexendoo** [comment]
Where inline suggestions become unwieldy you can replace `span_suggestion` with `span_suggestion_verbose`

If there is a suggestion it should contain the raw -> regular string conversion yeah. But it's also fine to not have a suggestion and just mention that it would have to be converted to a non-raw string

**Comment by ada4a** [comment]
> But it's also fine to not have a suggestion and just mention that it would have to be converted to a non-raw string

Yes, but I think both `non_ascii_literal` and _especially_ `invisible_characters` pretty much require a structured suggestion in order to be useful... One could of course create a `multipart_suggestion` for each and every linted character, but I'm afraid that the merging of the `MultiSpan` might result in the suggestion spanning pretty much the entire string, in which case the user will still see the rawness of the "suggested" literal, and be confused by it.

**Comment by shulaoda** [comment]
I ran into this too and opened #17302, which warns on raw strings instead of emitting a machine-applicable fix that corrupts them, since a wrong auto-fix is worse than no fix here.
